src/AggregateFunctions/AggregateFunctionSequenceNextNode.h

```diff
@@ -0,0 +1,462 @@
+
+        const UInt64 new_elems = std::min(data(rhs).value.size(), static_cast<size_t>(max_elems) - data(place).value.size());
+        for (UInt64 i = 0; i < new_elems; ++i)
+            a.push_back(b[i]->clone(arena), arena);
+
+        /// either sort whole container or do so partially merging ranges afterwards
+        using Comparator = typename SequenceNextNodeGeneralData<T, Node, Descending>::Comparator;
+
+        if (!data(place).sorted && !data(rhs).sorted)
+            std::stable_sort(std::begin(a), std::end(a), Comparator{});
+        else
+        {
+            const auto begin = std::begin(a);
+            const auto middle = std::next(begin, a_size);
+            const auto end = std::end(a);
+
+            if (!data(place).sorted)
+                std::stable_sort(begin, middle, Comparator{});
+
+            if (!data(rhs).sorted)
+                std::stable_sort(middle, end, Comparator{});
+
+            std::inplace_merge(begin, middle, end, Comparator{});
+        }
+
+        data(place).sorted = true;
+    }
+
+    void serialize(ConstAggregateDataPtr place, WriteBuffer & buf) const override
+    {
+        writeBinary(data(place).sorted, buf);
+
+        auto & value = data(place).value;
+        writeVarUInt(value.size(), buf);
+        for (auto & node : value)
+            node->write(buf);
+    }
+
+    void deserialize(AggregateDataPtr place, ReadBuffer & buf, Arena * arena) const override
+    {
+        readBinary(data(place).sorted, buf);
+
+        UInt64 size;
+        readVarUInt(size, buf);
+
+        if (unlikely(size == 0))
+            return;
+
+        auto & value = data(place).value;
+
+        value.resize(size, arena);
+        for (UInt64 i = 0; i < size; ++i)
+            value[i] = Node::read(buf, arena);
+    }
+
+    inline UInt32 calculateJump(const Data & data, const UInt32 i, const UInt32 j) const
+    {
+        UInt32 k = 0;
+        for (; k < events_size - j; ++k)
+            if (data.value[i - j]->events_bitset & (1 << (events_size - 1 - j - k)))
+                return k;
+        return k;
+    }
+
+    /// This method returns an index of next node that matched the events.
+    /// It is one as referring Boyer-Moore-Algorithm.
+    UInt32 getNextNodeIndex(Data & data) const
+    {
+        if (data.value.size() <= events_size)
+            return 0;
+
+        data.sort();
+
+        UInt32 i = events_size - 1;
+        while (i < data.value.size())
+        {
+            UInt32 j = 0;
+            /// It checks whether the chain of events are matched or not.
+            for (; j < events_size; ++j)
+                if (!(data.value[i - j]->events_bitset & (1 << (events_size - 1 - j))))
+                    break;
+
+            /// If the chain of events are matched returns the index of result value.
+            if (j == events_size)
+                return i + 1;
+
+            i += calculateJump(data, i, j);
+        }
+
+        return 0;
+    }
```

Comment:
This definitely require  verbose explanation about how algorithm works. If it is a part of other algorithm (Boyer-Moore) it need to be explained anyway, how it is used, and how is adapted.

Reply:
@achimbab 
Have you added comment about this (I suppose core) part?

Also I have another question: did you consider reusing functionality of other `sequence*` functions? What main difference, that requires separate algorithm?
https://github.com/ClickHouse/ClickHouse/blob/d9e22ba646d9946a4c3e3abee39a1842b552dfc3/src/AggregateFunctions/AggregateFunctionSequenceMatch.h

Reply:
@vdimir 
Sure, I added comment about this.
If it isn't enough, please let me know.
https://github.com/achimbab/ClickHouse/blob/pr_sequenceNextNode/src/AggregateFunctions/AggregateFunctionSequenceNextNode.h

Reply:
@vdimir 
There are some differences between two functions.
1. The return type
2. sequenceNextNode has to store values of event_column.
3. sequenceNextNode can find the index of result value in  event_column.

And two algorithmes have a functional difference. For example:
| dt | id | action |
| -- | -- | -------- |
| 1970-01-01 09:00:01 |  1 | a      |
| 1970-01-01 09:00:02 |  1 | b      |
| 1970-01-01 09:00:03 |  1 | b      |
| 1970-01-01 09:00:04 |  1 | c      |

```SQL
SELECT
    id,
    sequenceMatch('(?1)(?2)')(dt, action = 'a', action = 'c') AS result
FROM test_seq
GROUP BY id

 id   result
  1        1 
```
sequenceMatch returns 1. The order of events are `a -> b -> b -> c`, but It does not take into account event `b` in the middle, so `a -> c` makes sense even though there is `b` in the middle.

```SQL
SELECT
    id,
    sequenceNextNode(0)(dt, action, action = 'a', action = 'c') AS result
FROM test_seq
GROUP BY id

 id   result
  1   ᴺᵁᴸᴸ   
```
sequenceNextNode returns NULL. It is more sensitive in case where there is interrupting `b`.